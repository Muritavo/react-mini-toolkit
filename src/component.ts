import inquirer from 'inquirer';
import { mkdirSync, writeFileSync } from 'fs';
import { join } from 'path';

const l = require('debug')('react-mini-toolkit');

export const FEATURES = [{
    short: "Tests (@testing-library/react)",
    name: "tests",
    checked: true
}, {
    short: "SCSS (SCSS Modules)",
    name: "scss",
    checked: false
}, {
    short: "Storybook",
    name: "storybook",
    checked: false
}, {
    short: "Intl messages",
    name: "intl",
    checked: false
}] as const

export default function createComponent() {
    const rootFolder = process.env.INIT_CWD || './';
    l("Running at", rootFolder);
    inquirer.prompt([{
        type: "input",
        name: "name",
        message: "What is the name of the component?",
        validate: (str) => /[a-zA-Z]+/.test(str) ? true : "The component name must have only letters"
    }, {
        type: "input",
        name: "description",
        message: "Describe in a few words what is the objective of this component:"
    }, {
        type: "checkbox",
        name: "features",
        message: "Feature to be prebuilt",
        choices: FEATURES
    }]).then(({ name: componentName, description: componentDescription, features: generationFeatures }) => {
        l(generationFeatures)
        l("Creating the component", componentName);
        const componentFolder = join(rootFolder, componentName);
        mkdirSync(componentFolder, { recursive: true });
        writeFileSync(join(componentFolder, `index.tsx`), `export { default } from './${componentName}';\n`);
        writeFileSync(join(componentFolder, `${componentName}.tsx`), `import React from 'react';
${generationFeatures.includes('scss') ? `import Styles from './${componentName}.module.scss';\n` : ''}
/**
 * ${componentDescription}
 **/
export default function ${componentName}() {
    return <>${componentName}</>;
}`);
        if (generationFeatures.includes('tests'))
            writeFileSync(join(componentFolder, `${componentName}.test.tsx`), `import React from 'react';
import { render } from '@testing-library/react';
import Component from './${componentName}';

it('Should at least render :)', () => {
    render(<Component/>);
})`);
        if (generationFeatures.includes('scss'))
            writeFileSync(join(componentFolder, `${componentName}.module.scss`), ``);
    })
}

type Features = { [key in typeof FEATURES[number]['name']]: boolean };

export function generateOutput(componentName: string, componentDescription: string, features: Features) {
    const files: { [filename: string]: string } = {};

    files['index.tsx'] = `export { default } from './${componentName}';\n`;
    files[`${componentName}.tsx`] = generateMainFile(componentName, componentDescription, features);
    if (features.tests)
        files[`${componentName}.test.tsx`] = generateTestsFile(componentName, features);
    if (features.scss)
        files[`${componentName}.module.scss`] = ``;
    if (features.storybook)
        files[`${componentName}.stories.tsx`] = generateStorybookFile(componentName);
    if (features.intl)
        files[`messages.ts`] = generateReactIntlMessagesFile();

    return files;
}

export function generateReactIntlMessagesFile() {
    return `import { defineMessages } from 'react-intl';
    
export default defineMessages({})`;
}

export function replaceUpperCaseBySpace(componentName: string) {
    return componentName.replace(/(?=[A-Z])/g, " ").trim();
}

function generateStorybookFile(componentName: string) {
    return `import React from 'react';
import ${componentName} from "./${componentName}";
    
export default {
    component: ${componentName},
    title: "${replaceUpperCaseBySpace(componentName)}"
}

export const InitialImplementation = (...args: any) => <${componentName} {...args}/>;`;
}

function generateTestsFile(componentName: string, features: Features) {
    const whereToImportTheComponentFrom = features.storybook ? `import { InitialImplementation as Component } from './${componentName}.stories';` : `import Component from './${componentName}';`;
    const componentRender = features.intl ? `<IntlProvider locale="pt-br">
        <Component/>
    </IntlProvider>` : `<Component/>`
    return `import React from 'react';
import { render } from '@testing-library/react';
${features.intl ? "import { IntlProvider } from 'react-intl';" : ""}
${whereToImportTheComponentFrom}

it('Should at least render :)', () => {
    render(${componentRender});
})`;
}

function generateMainFile(componentName: string, componentDescription: string, features: { [key in typeof FEATURES[number]['name']]: boolean }) {
    return `import React from 'react';
${features.scss ? `import Styles from './${componentName}.module.scss';` : ''}
${features.intl ? `import { useIntl } from 'react-intl'
import messages from './messages'` : ''}

/**
 * ${componentDescription}
 **/
export default function ${componentName}() {
    ${features.intl ? "const { formatMessage } = useIntl();" : ""}
    return <>${componentName}</>;
}`;
}