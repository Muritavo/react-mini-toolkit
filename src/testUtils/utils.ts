import path from "path";
import * as ts from "typescript";

function workaroundForPathBecauseIDnotKnowTSCompilersYet(path: string) {
    if (path.startsWith("/node_modules"))
        return `.${path}`;
    else
        return path;
}

export function getDiagnosticsForText(files: {
    [filename: string]: string
}) {
    const fileNames = Object.keys(files);
    const asts = fileNames.reduce((r, filename) => ({
        ...r,
        [filename]: ts.createSourceFile(filename, files[filename], ts.ScriptTarget.ESNext)
    }), {}) as { [key: string]: ts.SourceFile }
    const options: ts.CompilerOptions = {
        target: ts.ScriptTarget.ESNext,
        module: ts.ModuleKind.CommonJS,
        moduleResolution: ts.ModuleResolutionKind.NodeJs,
        jsx: ts.JsxEmit.ReactJSX,
        esModuleInterop: true,
    };
    const nativeHost = ts.createCompilerHost(options);
    function fileIsMocked(filePath: string) {
        return fileNames.includes(filePath)
    }
    function cleanFilePath(f: string) {
        return f.replace(/^[^A-Za-z]/g, "");
    }
    const host: ts.CompilerHost = {
        fileExists: filePath => fileIsMocked(cleanFilePath(filePath)) || nativeHost.fileExists(workaroundForPathBecauseIDnotKnowTSCompilersYet(filePath)),
        directoryExists: dirPath => dirPath === "/" || nativeHost.directoryExists!(workaroundForPathBecauseIDnotKnowTSCompilersYet(dirPath)),
        getCurrentDirectory: () => "/",
        getDirectories: (s) => nativeHost.getDirectories!(s),
        getCanonicalFileName: fileName => fileName,
        getNewLine: () => "\n",
        getDefaultLibFileName: () => nativeHost.getDefaultLibFileName(options),
        getSourceFile: filePath => fileIsMocked(cleanFilePath(filePath)) ? asts[cleanFilePath(filePath)] : nativeHost.getSourceFile(workaroundForPathBecauseIDnotKnowTSCompilersYet(filePath), ts.ScriptTarget.Latest),
        readFile: filePath => fileIsMocked(cleanFilePath(filePath)) ? files[cleanFilePath(filePath)] : nativeHost.readFile(workaroundForPathBecauseIDnotKnowTSCompilersYet(filePath)),
        useCaseSensitiveFileNames: () => true,
        writeFile: () => { }
    };
    const program = ts.createProgram({
        options,
        rootNames: fileNames.concat(require.resolve("@types/jest/index.d.ts")),
        host
    });

    return ts.getPreEmitDiagnostics(program);
}