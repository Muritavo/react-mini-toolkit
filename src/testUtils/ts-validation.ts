import { getDiagnosticsForText } from "../testUtils/utils";

export function assertFilesCompile(outputFiles: {
    [fileName: string]: string
}) {
  const outputResult = getDiagnosticsForText(outputFiles).map(
    (o) => o.messageText
  );

  const errors = outputResult.filter((o) => {
    const a = JSON.stringify(o);
    return (
      !a.includes(`MockComponent.module.scss'`) &&
      !a.includes("README.md") &&
      !a.includes("'cypress'") &&
      !a.includes("'Cypress'") &&
      !a.includes("'JQuery'")
    );
  });
  expect(errors).toHaveLength(0);
}
