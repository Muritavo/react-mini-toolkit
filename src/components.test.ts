import { getDiagnosticsForText } from "./testUtils/utils";
import { FEATURES, generateOutput, replaceUpperCaseBySpace } from "./component";

const FEATURES_TO_TEST = FEATURES.map(f => f.name);

const boolCombo = (size: number) => {
    const buf = Array(1 << size)
    for (let i = buf.length; i--;) {
        buf[i] = Array(size)
        for (let j = size; j--;)
            buf[i][j] = !!+!!(i & 1 << j)
    }
    return buf
}

const allCombinations = boolCombo(FEATURES_TO_TEST.length);
const allFeatureCombinations = allCombinations.map((combination) => (FEATURES_TO_TEST.reduce((r, key, i) => ({ ...r, [key]: combination[i] }), {})));

it.each(allFeatureCombinations)("Features for the component generation when features are enabled as %s", (enabledFeatures) => {
    const outputFiles = generateOutput("MockComponent", "A component to check that everything is fine", enabledFeatures as any);
    expect(outputFiles).toMatchSnapshot();

    const outputResult = getDiagnosticsForText(outputFiles);
    const errors = outputResult.filter(o => !JSON.stringify(o.messageText).includes(`MockComponent.module.scss'`));
    expect(errors).toHaveLength(0);
})

it("Should replace with space", () => {
    expect(replaceUpperCaseBySpace("SomeComponent")).toEqual("Some Component");
})