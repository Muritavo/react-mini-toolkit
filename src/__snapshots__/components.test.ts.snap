// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Features for the component generation when features are enabled as { tests: false, scss: false, storybook: false, intl: false } 1`] = `
Object {
  "MockComponent.tsx": "import React from 'react';



/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
}
`;

exports[`Features for the component generation when features are enabled as { tests: false, scss: false, storybook: false, intl: true } 1`] = `
Object {
  "MockComponent.tsx": "import React from 'react';

import { useIntl } from 'react-intl'
import messages from './messages'

/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    const { formatMessage } = useIntl();
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
  "messages.ts": "import { defineMessages } from 'react-intl';
    
export default defineMessages({})",
}
`;

exports[`Features for the component generation when features are enabled as { tests: false, scss: false, storybook: true, intl: false } 1`] = `
Object {
  "MockComponent.stories.tsx": "import React from 'react';
import MockComponent from \\"./MockComponent\\";
    
export default {
    component: MockComponent,
    title: \\"Mock Component\\"
}

export const InitialImplementation = (...args: any) => <MockComponent {...args}/>;",
  "MockComponent.tsx": "import React from 'react';



/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
}
`;

exports[`Features for the component generation when features are enabled as { tests: false, scss: false, storybook: true, intl: true } 1`] = `
Object {
  "MockComponent.stories.tsx": "import React from 'react';
import MockComponent from \\"./MockComponent\\";
    
export default {
    component: MockComponent,
    title: \\"Mock Component\\"
}

export const InitialImplementation = (...args: any) => <MockComponent {...args}/>;",
  "MockComponent.tsx": "import React from 'react';

import { useIntl } from 'react-intl'
import messages from './messages'

/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    const { formatMessage } = useIntl();
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
  "messages.ts": "import { defineMessages } from 'react-intl';
    
export default defineMessages({})",
}
`;

exports[`Features for the component generation when features are enabled as { tests: false, scss: true, storybook: false, intl: false } 1`] = `
Object {
  "MockComponent.module.scss": "",
  "MockComponent.tsx": "import React from 'react';
import Styles from './MockComponent.module.scss';


/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
}
`;

exports[`Features for the component generation when features are enabled as { tests: false, scss: true, storybook: false, intl: true } 1`] = `
Object {
  "MockComponent.module.scss": "",
  "MockComponent.tsx": "import React from 'react';
import Styles from './MockComponent.module.scss';
import { useIntl } from 'react-intl'
import messages from './messages'

/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    const { formatMessage } = useIntl();
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
  "messages.ts": "import { defineMessages } from 'react-intl';
    
export default defineMessages({})",
}
`;

exports[`Features for the component generation when features are enabled as { tests: false, scss: true, storybook: true, intl: false } 1`] = `
Object {
  "MockComponent.module.scss": "",
  "MockComponent.stories.tsx": "import React from 'react';
import MockComponent from \\"./MockComponent\\";
    
export default {
    component: MockComponent,
    title: \\"Mock Component\\"
}

export const InitialImplementation = (...args: any) => <MockComponent {...args}/>;",
  "MockComponent.tsx": "import React from 'react';
import Styles from './MockComponent.module.scss';


/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
}
`;

exports[`Features for the component generation when features are enabled as { tests: false, scss: true, storybook: true, intl: true } 1`] = `
Object {
  "MockComponent.module.scss": "",
  "MockComponent.stories.tsx": "import React from 'react';
import MockComponent from \\"./MockComponent\\";
    
export default {
    component: MockComponent,
    title: \\"Mock Component\\"
}

export const InitialImplementation = (...args: any) => <MockComponent {...args}/>;",
  "MockComponent.tsx": "import React from 'react';
import Styles from './MockComponent.module.scss';
import { useIntl } from 'react-intl'
import messages from './messages'

/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    const { formatMessage } = useIntl();
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
  "messages.ts": "import { defineMessages } from 'react-intl';
    
export default defineMessages({})",
}
`;

exports[`Features for the component generation when features are enabled as { tests: true, scss: false, storybook: false, intl: false } 1`] = `
Object {
  "MockComponent.test.tsx": "import React from 'react';
import { render } from '@testing-library/react';

import Component from './MockComponent';

it('Should at least render :)', () => {
    render(<Component/>);
})",
  "MockComponent.tsx": "import React from 'react';



/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
}
`;

exports[`Features for the component generation when features are enabled as { tests: true, scss: false, storybook: false, intl: true } 1`] = `
Object {
  "MockComponent.test.tsx": "import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import Component from './MockComponent';

it('Should at least render :)', () => {
    render(<IntlProvider locale=\\"pt-br\\">
        <Component/>
    </IntlProvider>);
})",
  "MockComponent.tsx": "import React from 'react';

import { useIntl } from 'react-intl'
import messages from './messages'

/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    const { formatMessage } = useIntl();
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
  "messages.ts": "import { defineMessages } from 'react-intl';
    
export default defineMessages({})",
}
`;

exports[`Features for the component generation when features are enabled as { tests: true, scss: false, storybook: true, intl: false } 1`] = `
Object {
  "MockComponent.stories.tsx": "import React from 'react';
import MockComponent from \\"./MockComponent\\";
    
export default {
    component: MockComponent,
    title: \\"Mock Component\\"
}

export const InitialImplementation = (...args: any) => <MockComponent {...args}/>;",
  "MockComponent.test.tsx": "import React from 'react';
import { render } from '@testing-library/react';

import { InitialImplementation as Component } from './MockComponent.stories';

it('Should at least render :)', () => {
    render(<Component/>);
})",
  "MockComponent.tsx": "import React from 'react';



/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
}
`;

exports[`Features for the component generation when features are enabled as { tests: true, scss: false, storybook: true, intl: true } 1`] = `
Object {
  "MockComponent.stories.tsx": "import React from 'react';
import MockComponent from \\"./MockComponent\\";
    
export default {
    component: MockComponent,
    title: \\"Mock Component\\"
}

export const InitialImplementation = (...args: any) => <MockComponent {...args}/>;",
  "MockComponent.test.tsx": "import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { InitialImplementation as Component } from './MockComponent.stories';

it('Should at least render :)', () => {
    render(<IntlProvider locale=\\"pt-br\\">
        <Component/>
    </IntlProvider>);
})",
  "MockComponent.tsx": "import React from 'react';

import { useIntl } from 'react-intl'
import messages from './messages'

/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    const { formatMessage } = useIntl();
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
  "messages.ts": "import { defineMessages } from 'react-intl';
    
export default defineMessages({})",
}
`;

exports[`Features for the component generation when features are enabled as { tests: true, scss: true, storybook: false, intl: false } 1`] = `
Object {
  "MockComponent.module.scss": "",
  "MockComponent.test.tsx": "import React from 'react';
import { render } from '@testing-library/react';

import Component from './MockComponent';

it('Should at least render :)', () => {
    render(<Component/>);
})",
  "MockComponent.tsx": "import React from 'react';
import Styles from './MockComponent.module.scss';


/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
}
`;

exports[`Features for the component generation when features are enabled as { tests: true, scss: true, storybook: false, intl: true } 1`] = `
Object {
  "MockComponent.module.scss": "",
  "MockComponent.test.tsx": "import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import Component from './MockComponent';

it('Should at least render :)', () => {
    render(<IntlProvider locale=\\"pt-br\\">
        <Component/>
    </IntlProvider>);
})",
  "MockComponent.tsx": "import React from 'react';
import Styles from './MockComponent.module.scss';
import { useIntl } from 'react-intl'
import messages from './messages'

/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    const { formatMessage } = useIntl();
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
  "messages.ts": "import { defineMessages } from 'react-intl';
    
export default defineMessages({})",
}
`;

exports[`Features for the component generation when features are enabled as { tests: true, scss: true, storybook: true, intl: false } 1`] = `
Object {
  "MockComponent.module.scss": "",
  "MockComponent.stories.tsx": "import React from 'react';
import MockComponent from \\"./MockComponent\\";
    
export default {
    component: MockComponent,
    title: \\"Mock Component\\"
}

export const InitialImplementation = (...args: any) => <MockComponent {...args}/>;",
  "MockComponent.test.tsx": "import React from 'react';
import { render } from '@testing-library/react';

import { InitialImplementation as Component } from './MockComponent.stories';

it('Should at least render :)', () => {
    render(<Component/>);
})",
  "MockComponent.tsx": "import React from 'react';
import Styles from './MockComponent.module.scss';


/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
}
`;

exports[`Features for the component generation when features are enabled as { tests: true, scss: true, storybook: true, intl: true } 1`] = `
Object {
  "MockComponent.module.scss": "",
  "MockComponent.stories.tsx": "import React from 'react';
import MockComponent from \\"./MockComponent\\";
    
export default {
    component: MockComponent,
    title: \\"Mock Component\\"
}

export const InitialImplementation = (...args: any) => <MockComponent {...args}/>;",
  "MockComponent.test.tsx": "import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { InitialImplementation as Component } from './MockComponent.stories';

it('Should at least render :)', () => {
    render(<IntlProvider locale=\\"pt-br\\">
        <Component/>
    </IntlProvider>);
})",
  "MockComponent.tsx": "import React from 'react';
import Styles from './MockComponent.module.scss';
import { useIntl } from 'react-intl'
import messages from './messages'

/**
 * A component to check that everything is fine
 **/
export default function MockComponent() {
    const { formatMessage } = useIntl();
    return <>MockComponent</>;
}",
  "index.tsx": "export { default } from './MockComponent';
",
  "messages.ts": "import { defineMessages } from 'react-intl';
    
export default defineMessages({})",
}
`;
